R/synthdid.R

```diff
@@ -179,21 +184,67 @@ synthdid_effect_curve = function(estimate) {
+#'        Defaults to NULL, which indicates that it should be the fraction of post-treatment to pre-treatment data
 #' @export synthdid_se
-synthdid_se = function(estimate, weights = attr(estimate, 'weights')) {
+synthdid_se = function(estimate,
+                       method = c("jackknife", "bootstrap", "placebo"),
+                       weights = attr(estimate, 'weights'),
+                       repetitions = 25,
+                       treated.fraction = NULL) {
+  method = match.arg(method)
   setup = attr(estimate, 'setup')
   opts = attr(estimate, 'opts')
+  estimator = attr(estimate, 'estimator')
   if (setup$N0 == nrow(setup$Y) - 1) { return(NA) }
 
-  sum_normalize = function(x) { x / sum(x) }
-  theta = function(ind) {
-    weights.jk = weights
-    if (!is.null(weights)) { weights.jk$omega = sum_normalize(weights$omega[ind[ind <= setup$N0]]) }
-    estimate.jk = synthdid_estimate(setup$Y[ind, ], sum(ind <= setup$N0), setup$T0, X = setup$X[ind, , ],
-      zeta.lambda = opts$zeta.lambda, zeta.omega = opts$zeta.omega,
-      lambda.intercept = opts$lambda.intercept, omega.intercept = opts$omega.intercept,
-      weights = weights.jk)
+  if (method == "jackknife") {
+    sum_normalize = function(x) { x / sum(x) }
+    theta = function(ind) {
+      weights.jk = weights
+      if (!is.null(weights)) { weights.jk$omega = sum_normalize(weights$omega[ind[ind <= setup$N0]]) }
+      estimate.jk = synthdid_estimate(setup$Y[ind, ], sum(ind <= setup$N0), setup$T0, X = setup$X[ind, , ],
+        zeta.lambda = opts$zeta.lambda, zeta.omega = opts$zeta.omega,
+        lambda.intercept = opts$lambda.intercept, omega.intercept = opts$omega.intercept,
+        weights = weights.jk)
+    }
+    return (jackknife(1:nrow(setup$Y), theta))
+  } else if (method == "bootstrap") {
+    setup.bs = setup
+    treatment_ind = c(rep(0,setup$N0),rep(1,nrow(setup$Y) - setup$N0))
+    theta = function(ind) {
+      Y_init = setup$Y[ind,]
+      X_init = setup$X[ind, , ]
+      tr_stat = treatment_ind[ind]
+      tr_index = order(tr_stat)
+      Y_b = Y_init[tr_index,]
+      X_b = X_init[tr_index, , ]
+      n_0 = nrow(setup$Y)-sum(tr_stat)
+      if (sum(tr_stat) == 0 || sum(tr_stat) == nrow(setup$Y)) {
+        return (NA)
+      }
+      setup.bs$Y = Y_b
+      setup.bs$N0 = n_0
+      setup.bs$X = X_b
+      do.call(estimator, c(setup.bs, opts))
+    }
+    return (bootstrap(repetitions, nrow(setup$Y), nrow(setup$Y), estimate, theta))
+  } else if (method == "placebo") {
+    if (is.null(treated.fraction)) { treated.fraction = 1 - setup$T0 / ncol(setup$Y) }
+    n1 = floor(nrow(setup$Y) * treated.fraction)
+    setup.bs = setup
+    theta = function(ind) {
+      Y_b = rbind(setup$Y[-ind,], setup$Y[ind,])
+      # X_b = rbind(setup$X[-ind, ,], setup$X[ind, ,])
+      n_0 = nrow(setup$Y) - n1
+      setup.bs$Y = Y_b
+      setup.bs$N0 = n_0
+      # setup.bs$X = X_b
```

Comment:
Why are covariates commented out for the bootstrap?

Reply:
Pending an rbind-workaround for a 3d array, suggestion for that?

Reply:
cat